File: multi-processor-programming-study/exercises/cpp/part2/counting/exercise01_parallel_counting.cpp
/**
 * Exercise: Parallel Counting (C++20)
 *
 * CONCEPT: Efficient parallel counting using various strategies
 *
 * Strategies:
 * - Cache-line padding to avoid false sharing
 * - Thread-local aggregation
 * - Striping across multiple counters
 *
 * LEARNING OBJECTIVES:
 * - Understand false sharing and cache lines
 * - Implement scalable counters in C++
 * - Use thread_local storage
 * - Compare performance under contention
 */

#include <iostream>
#include <thread>
#include <vector>
#include <atomic>
#include <chrono>
#include <iomanip>
#include <numeric>

namespace multiprocessor::part2::counting {

/**
 * TODO: Implement Cache-Line Padded Counter
 *
 * Avoid false sharing by padding to cache line size (64 bytes)
 */
class alignas(64) PaddedCounter {
private:
    std::atomic<long> value{0};
    // Padding to fill cache line
    char padding[64 - sizeof(std::atomic<long>)];

public:
    void increment() {
        value.fetch_add(1, std::memory_order_relaxed);
    }

    long get() const {
        return value.load(std::memory_order_relaxed);
    }
};

/**
 * TODO: Implement Striped Counter
 *
 * Distribute count across multiple counters to reduce contention
 */
class StripedCounter {
private:
    std::vector<PaddedCounter> stripes;

public:
    StripedCounter(int num_stripes) : stripes(num_stripes) {}

    /**
     * TODO: Implement increment
     *
     * Hash thread ID to stripe to reduce contention
     */
    void increment() {
        // TODO: Get thread ID and hash to stripe
        size_t thread_id = std::hash<std::thread::id>{}(std::this_thread::get_id());
        size_t stripe_id = thread_id % stripes.size();
        stripes[stripe_id].increment();
    }

    /**
     * TODO: Implement get
     */
    long get() const {
        // TODO: Sum all stripes
        long sum = 0;
        for (const auto& stripe : stripes) {
            sum += stripe.get();
        }
        return sum;
    }
};

/**
 * TODO: Implement Thread-Local Counter with Combining
 */
class CombiningCounter {
private:
    std::atomic<long> global_counter{0};
    static constexpr int COMBINING_THRESHOLD = 100;

    struct ThreadLocalData {
        long local_count = 0;
    };

    static thread_local ThreadLocalData tl_data;

public:
    /**
     * TODO: Implement increment with combining
     */
    void increment() {
        // TODO: Increment thread-local counter
        tl_data.local_count++;

        // TODO: When threshold reached, flush to global
        if (tl_data.local_count >= COMBINING_THRESHOLD) {
            global_counter.fetch_add(tl_data.local_count, std::memory_order_relaxed);
            tl_data.local_count = 0;
        }
    }

    long get() const {
        // Note: This doesn't account for unflushed thread-local counts
        // In production, would need to iterate all thread-local data
        return global_counter.load(std::memory_order_relaxed);
    }
};

thread_local CombiningCounter::ThreadLocalData CombiningCounter::tl_data;

/**
 * Performance testing framework
 */
class PerformanceTest {
public:
    template <typename Counter>
    static long test_counter(const std::string& name, Counter& counter,
                            int num_threads, int increments_per_thread) {
        std::vector<std::thread> threads;

        auto start_time = std::chrono::high_resolution_clock::now();

        for (int i = 0; i < num_threads; ++i) {
            threads.emplace_back([&counter, increments_per_thread]() {
                for (int j = 0; j < increments_per_thread; ++j) {
                    counter.increment();
                }
            });
        }

        for (auto& t : threads) {
            t.join();
        }

        auto end_time = std::chrono::high_resolution_clock::now();
        auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(
            end_time - start_time).count();

        std::cout << std::left << std::setw(25) << name << ": "
                  << std::right << std::setw(6) << duration << "ms"
                  << " (count=" << counter.get() << ")\n";

        return duration;
    }
};

void test_correctness() {
    const int num_threads = 10;
    const int increments_per_thread = 1000;
    const long expected = num_threads * increments_per_thread;

    std::cout << "Testing correctness (threads=" << num_threads
              << ", increments=" << increments_per_thread << "):\n\n";

    // Test Striped Counter
    StripedCounter striped(num_threads);
    std::vector<std::thread> threads;
    for (int i = 0; i < num_threads; ++i) {
        threads.emplace_back([&striped, increments_per_thread]() {
            for (int j = 0; j < increments_per_thread; ++j) {
                striped.increment();
            }
        });
    }
    for (auto& t : threads) t.join();

    std::cout << "Striped Counter: Expected=" << expected
              << ", Actual=" << striped.get()
              << (striped.get() == expected ? " ✅\n" : " ❌\n");
}

void compare_performance() {
    const int num_threads = 16;
    const int increments_per_thread = 100000;

    std::cout << "\n=== Performance Comparison ===\n";
    std::cout << "High contention test (threads=" << num_threads
              << ", increments=" << increments_per_thread << "):\n\n";

    // Baseline: Single atomic counter
    std::atomic<long> atomic_counter{0};
    PerformanceTest::test_counter("std::atomic<long>",
        [&](){ atomic_counter.fetch_add(1); return atomic_counter.load(); },
        num_threads, increments_per_thread);

    // Padded Counter (array of one)
    std::vector<PaddedCounter> padded_array(1);
    auto padded_inc = [&padded_array]() { padded_array[0].increment(); };
    auto padded_get = [&padded_array]() { return padded_array[0].get(); };

    std::vector<std::thread> threads;
    auto start = std::chrono::high_resolution_clock::now();
    for (int i = 0; i < num_threads; ++i) {
        threads.emplace_back([&padded_inc, increments_per_thread]() {
            for (int j = 0; j < increments_per_thread; ++j) {
                padded_inc();
            }
        });
    }
    for (auto& t : threads) t.join();
    auto padded_time = std::chrono::duration_cast<std::chrono::milliseconds>(
        std::chrono::high_resolution_clock::now() - start).count();
    std::cout << std::left << std::setw(25) << "Padded Counter" << ": "
              << std::right << std::setw(6) << padded_time << "ms"
              << " (count=" << padded_get() << ")\n";

    // Striped Counter
    StripedCounter striped(num_threads);
    PerformanceTest::test_counter("Striped Counter", striped,
        num_threads, increments_per_thread);

    std::cout << "\n💡 Striping reduces contention by distributing load\n";
    std::cout << "💡 Padding prevents false sharing between cache lines\n";
}

} // namespace multiprocessor::part2::counting

int main() {
    using namespace multiprocessor::part2::counting;

    std::cout << "=== Parallel Counting Test (C++) ===\n\n";

    test_correctness();
    compare_performance();

    return 0;
}

File: multi-processor-programming-study/exercises/java/src/main/java/com/multiprocessor/part2/counting/Exercise01_ParallelCounting.java
package com.multiprocessor.part2.counting;

import java.util.concurrent.atomic.AtomicLong;
import java.util.concurrent.atomic.LongAdder;
import java.util.Arrays;
import java.util.concurrent.ForkJoinPool;
import java.util.concurrent.RecursiveTask;

/**
 * Exercise: Parallel Counting and Reduction
 *
 * CONCEPT: Efficient parallel counting using various strategies
 *
 * Strategies:
 * - Combining: Threads combine their partial results
 * - Striping: Distribute counts across cache lines
 * - Tree-based: Hierarchical reduction
 *
 * LEARNING OBJECTIVES:
 * - Implement scalable counters
 * - Use LongAdder for high contention
 * - Understand false sharing and cache line padding
 * - Implement parallel reduction patterns
 */
public class Exercise01_ParallelCounting {

    /**
     * TODO: Implement Combining Counter
     *
     * Threads combine their increments before accessing shared counter
     */
    public static class CombiningCounter {
        private final AtomicLong counter = new AtomicLong(0);
        private static final int COMBINING_THRESHOLD = 100;

        // Thread-local buffer for combining
        private final ThreadLocal<Long> localBuffer = ThreadLocal.withInitial(() -> 0L);

        /**
         * TODO: Implement increment with combining
         */
        public void increment() {
            // TODO: Increment thread-local buffer
            // TODO: When buffer reaches threshold, flush to global counter
        }

        /**
         * TODO: Implement flush
         */
        private void flush() {
            // TODO: Add local buffer to global counter atomically
            // TODO: Reset local buffer
        }

        public long get() {
            // TODO: Sum global counter + all pending local buffers
            return counter.get();
        }
    }

    /**
     * TODO: Implement Cache-line Padded Counter
     *
     * Avoid false sharing by padding to cache line size
     */
    public static class PaddedCounter {
        // Cache line is typically 64 bytes
        // Java object header is ~16 bytes
        // long is 8 bytes
        // Need ~40 bytes padding

        private volatile long p0, p1, p2, p3, p4; // Padding before
        private volatile long value = 0;
        private volatile long p5, p6, p7, p8, p9; // Padding after

        /**
         * TODO: Implement thread-safe increment
         */
        public synchronized void increment() {
            // TODO: Increment value
        }

        public long get() {
            return value;
        }
    }

    /**
     * TODO: Implement Parallel Array Sum using Fork/Join
     *
     * Divide-and-conquer parallel reduction
     */
    public static class ParallelArraySum extends RecursiveTask<Long> {
        private static final int THRESHOLD = 1000;
        private final long[] array;
        private final int start;
        private final int end;

        public ParallelArraySum(long[] array, int start, int end) {
            this.array = array;
            this.start = start;
            this.end = end;
        }

        @Override
        protected Long compute() {
            // TODO: If range small enough, compute sequentially
            // TODO: Otherwise, split in half and fork both subtasks
            // TODO: Join results from both subtasks
            return 0L;
        }

        public static long sum(long[] array) {
            // TODO: Create ForkJoinPool and invoke task
            return 0L;
        }
    }

    /**
     * TODO: Implement Scalable Counter using Striping
     *
     * Distribute count across multiple counters to reduce contention
     */
    public static class StripedCounter {
        private final AtomicLong[] stripes;
        private final int numStripes;

        public StripedCounter(int numStripes) {
            this.numStripes = numStripes;
            this.stripes = new AtomicLong[numStripes];
            // TODO: Initialize each stripe
            for (int i = 0; i < numStripes; i++) {
                stripes[i] = new AtomicLong(0);
            }
        }

        /**
         * TODO: Implement increment
         *
         * Hash thread ID to stripe to reduce contention
         */
        public void increment() {
            // TODO: Get thread ID hash
            // TODO: Select stripe based on hash
            // TODO: Increment selected stripe
        }

        /**
         * TODO: Implement get
         */
        public long get() {
            // TODO: Sum all stripes
            long sum = 0;
            for (AtomicLong stripe : stripes) {
                sum += stripe.get();
            }
            return sum;
        }
    }

    /**
     * Performance comparison
     */
    static class PerformanceTest {
        static void testCounter(String name, Runnable incrementOp,
                               java.util.function.LongSupplier getOp,
                               int numThreads, int incrementsPerThread)
                throws InterruptedException {

            Thread[] threads = new Thread[numThreads];
            long startTime = System.nanoTime();

            for (int i = 0; i < numThreads; i++) {
                threads[i] = new Thread(() -> {
                    for (int j = 0; j < incrementsPerThread; j++) {
                        incrementOp.run();
                    }
                });
                threads[i].start();
            }

            for (Thread t : threads) {
                t.join();
            }

            long endTime = System.nanoTime();
            long duration = (endTime - startTime) / 1_000_000; // ms

            System.out.printf("%-20s: %6dms (count=%d)%n",
                            name, duration, getOp.getAsLong());
        }
    }

    /**
     * Demonstration and testing
     */
    public static void main(String[] args) throws InterruptedException {
        System.out.println("=== Parallel Counting Test ===\n");

        testCorrectness();
        System.out.println("\n=== Performance Comparison ===");
        comparePerformance();
        System.out.println("\n=== Parallel Array Sum ===");
        testParallelSum();
    }

    private static void testCorrectness() throws InterruptedException {
        int numThreads = 10;
        int incrementsPerThread = 1000;
        int expected = numThreads * incrementsPerThread;

        System.out.println("Testing correctness (threads=" + numThreads +
                         ", increments=" + incrementsPerThread + "):\n");

        // Test Striped Counter
        StripedCounter striped = new StripedCounter(numThreads);
        Thread[] threads = new Thread[numThreads];
        for (int i = 0; i < numThreads; i++) {
            threads[i] = new Thread(() -> {
                for (int j = 0; j < incrementsPerThread; j++) {
                    striped.increment();
                }
            });
            threads[i].start();
        }
        for (Thread t : threads) t.join();

        System.out.println("Striped Counter: Expected=" + expected +
                         ", Actual=" + striped.get() +
                         (striped.get() == expected ? " ✅" : " ❌"));

        // Test LongAdder (Java's optimized counter)
        LongAdder adder = new LongAdder();
        threads = new Thread[numThreads];
        for (int i = 0; i < numThreads; i++) {
            threads[i] = new Thread(() -> {
                for (int j = 0; j < incrementsPerThread; j++) {
                    adder.increment();
                }
            });
            threads[i].start();
        }
        for (Thread t : threads) t.join();

        System.out.println("LongAdder:       Expected=" + expected +
                         ", Actual=" + adder.sum() +
                         (adder.sum() == expected ? " ✅" : " ❌"));
    }

    private static void comparePerformance() throws InterruptedException {
        int numThreads = 16;
        int incrementsPerThread = 100000;

        System.out.println("High contention test (threads=" + numThreads +
                         ", increments=" + incrementsPerThread + "):\n");

        // AtomicLong (baseline)
        AtomicLong atomicCounter = new AtomicLong(0);
        PerformanceTest.testCounter("AtomicLong",
            atomicCounter::incrementAndGet,
            atomicCounter::get,
            numThreads, incrementsPerThread);

        // Striped Counter
        StripedCounter striped = new StripedCounter(numThreads);
        PerformanceTest.testCounter("StripedCounter",
            striped::increment,
            striped::get,
            numThreads, incrementsPerThread);

        // LongAdder
        LongAdder adder = new LongAdder();
        PerformanceTest.testCounter("LongAdder",
            adder::increment,
            adder::sum,
            numThreads, incrementsPerThread);

        System.out.println("\n💡 LongAdder uses dynamic striping and is highly optimized");
        System.out.println("💡 StripedCounter shows the principle behind LongAdder");
    }

    private static void testParallelSum() {
        int size = 10_000_000;
        long[] array = new long[size];
        Arrays.fill(array, 1);

        System.out.println("Array size: " + size);

        // Sequential
        long start = System.nanoTime();
        long seqSum = 0;
        for (long value : array) {
            seqSum += value;
        }
        long seqTime = (System.nanoTime() - start) / 1_000_000;

        // Parallel
        start = System.nanoTime();
        long parSum = ParallelArraySum.sum(array);
        long parTime = (System.nanoTime() - start) / 1_000_000;

        System.out.println("Sequential sum: " + seqSum + " (" + seqTime + "ms)");
        System.out.println("Parallel sum:   " + parSum + " (" + parTime + "ms)");
        System.out.println("Speedup: " + String.format("%.2fx", (double)seqTime / parTime));
        System.out.println(seqSum == parSum ? "✅ Correct" : "❌ Incorrect");
    }
}
